### DISCUSSION TITLE
Zmk mouse behaviors

### DISCUSSION THREAD

**Discussion by valdar** [nickcoutsos/keymap-editor#148] [Ideas]
Hello would you be interested in adding zmk mouse behaviors? I know that is an experimental feature still not merged in the main branch but would be nice nevertheless. 
I can provide the json descriptors to be added to `zmk-behaveirs.json` file if needed.

Someone else mentioned this as well in the third item of discurrion https://github.com/nickcoutsos/keymap-editor/discussions/62

**Comment by nickcoutsos** [comment]
Hey, yeah I would be interested but it'd be a little different than the `zmk-behaviors.json`.

There are a number of features currently in PR that I'd like to be able to add support for both because:

1. They're in high demand despite whatever challenges are delaying them
2. It'd be nice to be do more work up-front to add features

To that end I've been thinking of adding a new collection of behavior data that includes some high level information about which ZMK PR/fork enables it, and then including options in the settings menu where a user can enable them along with a clear disclaimer that the keymap editor won't do anything to their build config for them. Once the selected features are enabled they'd appear as normal behaviors in the list.

None of these is in place yet, but the mouse stuff seems like a good example. What would help me is knowing if there's a de facto standard mouse implementation so that I don't have conflicting behaviors. I haven't actually tried any of the various pointing device implementations so I'm not totally sure what each offers.

**Comment by valdar** [comment]
> None of these is in place yet, but the mouse stuff seems like a good example. What would help me is knowing if there's a de facto standard mouse implementation so that I don't have conflicting behaviors. I haven't actually tried any of the various pointing device implementations so I'm not totally sure what each offers.

The one I use follow from this work https://github.com/zmkfirmware/zmk/pull/778 in particular this one https://github.com/ftc/zmk/tree/mouse-ftc for Zephyr < 3.2 and this one https://github.com/urob/zmk/tree/mouse-3.2 for Zephyr 3.2

**Reply by nickcoutsos** [reply]
It's not available yet, but I've put together the schemas to enable this. If I find enough time tonight I'll take a stab at exposing the feature flags in the UI and will update here.

**Reply by nickcoutsos** [reply]
@valdar you can now enable mouse behaviors in the settings dialog. Have a look when you have a chance and let me know how it goes!

**Reply by valdar** [reply]
In the meantime the mouse support has been merged! https://github.com/zmkfirmware/zmk/pull/1991

**Reply by willnapier** [reply]
Hi Nick, firstly thank you very much for your excellent and helpful work on this editor. I am trying to get my first custom layout working on my Corne keyboard. I get errors every time so far similar to those that another user here mentioned. I would like to understand whether this is due to my mouse key layout implementation (actually I'm sure it is). I have enabled support in the editor but I am probably getting the syntax wrong. I selected mouse move behaviour but then don't really have a clue what to do next. I tried 'MouseMove' as a name/category but I see from Valdar's nav layer below that the code is &mmv. How do I populate the edit behaviour window with 'binding' etc and what would good defaults be for ms delay etc? With thanks, Will

**Reply by nickcoutsos** [reply]
Hi @willnapier,

I took a quick peek at your repo (https://github.com/willnapier/zmk-config-corne) and it looks like you're building against the main ZMK branch but they haven't merged in mouse movement behaviors yet.

The error you're seeing:
```
devicetree error: /tmp/zmk-config/config/corne.keymap:224 (column 61): parse error: expected number or parenthesized expression
```
is because the `MOVE_LEFT` and other names aren't defined (neither would the mouse move behavior but this particular failure comes first). To build against a fork that supports mouse movement you'd change `config/west.yml` to look something like:

```yaml
manifest:
  remotes:
    - name: zmkfirmware
      url-base: https://github.com/zmkfirmware
    - name: petejohanson
      url-base: https://github.com/petejohanson
  projects:
    - name: zmk
      remote: petejohanson
      revision: feat/pointers-move-scroll
      import: app/west.yml
  self:
    path: config
```

Once that's in place you should be able to use the default `&mmv` behavior instead of defining your own.

**Comment by valdar** [comment]
![image](https://github.com/nickcoutsos/keymap-editor/assets/8052/5488130a-ccc2-4587-a88f-3a0d279394e9)

Works likes a charm tnx! It might be nice to have some icons that better fit the keyspace and easily identifies mouse actions.

**Reply by nickcoutsos** [reply]
Icons would make sense, though I'm currently limited to the FontAwesome free collection so it wouldn't very clearly distinguish between different mouse behaviors or parameters. Eventually I'll need to switch to a different icon set and/or maybe come up with an "icon composition" system to let me add detail easily.

I guess [mouse] + [arrow] is clear enough for movement, but what would you expect for mouse scrolling?

**Reply by valdar** [reply]
mmm I guess `<i class="fa-thin fa-scroll"></i>` + [arrow] ? Little bit of a stretch but might work waiting for the icon composition system.

**Reply by nickcoutsos** [reply]
By composition I do mean specifying two or more icons (possibly also how they might be arranged but maybe that's too much) so not something I can do right now anyway.

**Comment by 6995050** [comment]
Hi! Many people get benefits from nickcoutsos easy remap app in github...that means not many of us are familiar with merging...is there any tutorial for noobie like me to map mousekey? (created a new branch, added some mousekey behavior, but failed to build it, and i believe we need some merging of github instead of tick the setting and remap the keys?). Appreciate your help! 

```
-- Configuring incomplete, errors occurred!
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake -DWEST_PYTHON=/usr/bin/python3 -B/tmp/tmp.oNQ3bUemLC -GNinja -DBOARD=nice_nano_v2 -DZMK_CONFIG=/__w/corne-wireless-view-zmk-config/corne-wireless-view-zmk-config/config '-DSHIELD=corne_left nice_view_adapter nice_view' -S/__w/corne-wireless-view-zmk-config/corne-wireless-view-zmk-config/zmk/app
Error: Process completed with exit code 1.
```

**Reply by nickcoutsos** [reply]
Hi there!

This is something I specifically want to avoid trying to automate because there are so many variables to consider. As far as merging that's only something you'd need to do if you want to combine multiple features not-yet implemented in the main ZMK repo. If all you're looking for is the mouse emulation stuff there are a number of branches maintained by ZMK users for this, I think a popular choice as @valdar points out is https://github.com/urob/zmk/tree/mouse-3.2

If you want to build from that you can update your `config/west.yml` to the following

```yml
manifest:
  remotes:
    - name: zmkfirmware
      url-base: https://github.com/zmkfirmware
    - name: urob
      url-base: https://github.com/urob
  projects:
    - name: zmk
      remote: urob
      revision: mouse-3.2
      import: app/west.yml
  self:
    path: config
```

Note that I haven't done this myself, but if you're interested in the details they're documented on the [ZMK Docs: Beta Testing](https://zmk.dev/docs/features/beta-testing) page.

**Comment by willnapier** [comment]
Hi Nick

Thank you for responding so quickly and helpfully! I am not a developer but
I have developed a thing about split ergo keyboards, hrms, layouts. I have
been using mouse keys on my Ultimate Hacking Keyboard for several years and
it is in my muscle memory to the extent that I couldn't do without mouse
movement, so your tool and its compatibilty with the branch that enables
mouse movement could not be more important to me. Much appreciated.

So can you confirm the branch I should be building against? Is it
github.com/urob/zmk? Or is it the petejohanson url?

Also, is the change to the config/west.yml the only thing I need to change?

With many thanks, Will

**Reply by nickcoutsos** [reply]
Oh sorry, I didn't notice that I had both examples in this discussion, I think either should work but I haven't kept up with them to see what's changed. Maybe try the petejohanson one first and if that doesn't work, try urob's.
